akka-projection-core/src/main/scala/akka/projection/Projection.scala

```diff
@@ -41,16 +42,38 @@ trait Projection[Envelope] {
   /**
    * Run the Projection.
    */
-  def run()(implicit systemProvider: ClassicActorSystemProvider): Unit
+  def run()(implicit systemProvider: ClassicActorSystemProvider): RunningProjection
 
-  def runWithBackoff()(implicit systemProvider: ClassicActorSystemProvider): Unit
+  def runWithBackoff()(implicit systemProvider: ClassicActorSystemProvider): RunningProjection
+
+  def withSettings(projectionSettings: ProjectionSettings): Projection[Envelope]
+
+}
+
+trait ProjectionState {
```

Comment:
I'm not sure I see why this is needed in `core`?

Reply:
We need an intermediate state because of the testkit. We need to be able to return the mappedSource without running it. There are certainly other ways of doing it. It can be an responsibility of the projection impl to make sure there is some manages state that is not leaked.

Reply:
testkit is calling `projection.mappedSource()`, not this state.
I got the impression that it's only enforcing a specific way of how to implement Projection/RunningProjection, which is not needed at this point.

Reply:
`mappedSource()` depends on state because of the kill switch that's why it's all moved inside ProjectionState impl.

Reply:
`projection.mappedSource()` init the state and call `projectionState.mappedSource()`.